Indent error with the second function argument with eslint v1.3.1

Similar error was reported as issue #3591 which has been fixed since commit e8081 (I've confirmed), but I am still seeing the similar error with following code using rev 52c42b:

```js
'use strict';
/*eslint-disable no-console */
var Promise = require('bluebird');

function doSomething() {
    return Promise.resolve()
    .then(function () {
        var o = 1 + 2;
        return o;
    }, function (err) {
        console.log('err:', err);
        var o = 1 - 2;
        return o;
    });
}

doSomething()
.then(function (ans) {
    console.log('ans=%d', ans);
});
```
Here's the output:
```
test.js|11 col 9 error| Expected indentation of 12 space characters but found 8. (indent)
test.js|12 col 9 error| Expected indentation of 12 space characters but found 8. (indent)
test.js|13 col 9 error| Expected indentation of 12 space characters but found 8. (indent)
test.js|14 col 5 error| Expected indentation of 8 space characters but found 4. (indent)
```
My indent rule is `[2, 4]`

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/26349191-indent-error-with-the-second-function-argument-with-eslint-v1-3-1?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

Here is another example:

```js
Promise.all([
    promise1,
    promise2
]).then(function () {
    return promise3();
}).then(function () {
    // do something
}, function (e) {
    console.error(e);
});
```

```
promise.js
   9:5  error  Expected indentation of 8 space characters but found 4  indent
  10:1  error  Expected indentation of 4 space characters but found 0  indent
```

--- a/lib/cli-engine.js
+++ b/lib/cli-engine.js
@@ -475,8 +475,8 @@ CLIEngine.prototype = {
      */
     resolveFileGlobPatterns: function(patterns) {
         var extensions = this.options.extensions.map(function(ext) {
-                return ext.charAt(0) === "." ? ext.substr(1) : ext;
-            });
+            return ext.charAt(0) === "." ? ext.substr(1) : ext;
+        });
 
         return patterns.map(processPath(extensions));
     },

--- a/lib/rules/id-length.js
+++ b/lib/rules/id-length.js
@@ -19,10 +19,10 @@ module.exports = function(context) {
     var properties = options.properties !== "never";
     var exceptions = (options.exceptions ? options.exceptions : [])
         .reduce(function(obj, item) {
-        obj[item] = true;
+            obj[item] = true;
 
-        return obj;
-    }, {});
+            return obj;
+        }, {});
 
     var SUPPORTED_EXPRESSIONS = {
         "MemberExpression": function(parent) {

--- a/lib/rules/indent.js
+++ b/lib/rules/indent.js
@@ -255,86 +255,6 @@ module.exports = function(context) {
     }
 
     /**
-     * Check indent for function block content
-     * @param {ASTNode} node node to examine
-     * @returns {void}
-     */
-    function checkIndentInFunctionBlock(node) {
-
-        // Search first caller in chain.
-        // Ex.:
-        //
-        // Models <- Identifier
-        //   .User
-        //   .find()
-        //   .exec(function() {
-        //   // function body
-        // });
-        //
-        // Looks for 'Models'
-        var calleeNode = node.parent; // FunctionExpression
-        while (calleeNode.parent &&
-        calleeNode.parent.type === "CallExpression") {
-            calleeNode = calleeNode.parent;
-        }
-
-        var indent;
-
-        if (calleeNode.parent &&
-            (calleeNode.parent.type === "Property" ||
-            calleeNode.parent.type === "ArrayExpression")) {
-            // If function is part of array or object, comma can be put at left
-            indent = getNodeIndent(calleeNode, false, true);
-        } else {
-            // If function is standalone, simple calculate indent
-            indent = getNodeIndent(calleeNode);
-        }
-
-        // If part of a multi-line chain call, indent one more to differentiate:
-        // var a = new P()
-        //     .done(function (result) {
-        //         doUsefulStuff(result);
-        //     });
-        var calleeLoc = calleeNode.type === "CallExpression" && calleeNode.callee.loc;
-        if (calleeLoc &&
-            calleeLoc.end.line - calleeLoc.start.line === 1 &&
-            calleeLoc.end.line < node.loc.start.line
-        ) {
-            indent += indentSize;
-        }
-
-        indent += indentSize;
-        // If function content is not empty
-        var bodyIndent;
-        if (node.body.length > 0) {
-            bodyIndent = getNodeIndent(node.body[0]);
-            // Calculate left shift position don't require strict indent
-            // allow function body align to (indentSize * X)
-            while (bodyIndent > indent) {
-                indent += indentSize;
-            }
-
-            checkNodesIndent(node.body, indent);
-        }
-
-        checkLastNodeLineIndent(node, indent - indentSize);
-    }
-
-
-    /**
-     * Checks if the given node starts and ends on the same line
-     * @param {ASTNode} node The node to check
-     * @returns {boolean} Whether or not the block starts and ends on the same line.
-     */
-    function isSingleLineNode(node) {
-        var lastToken = context.getLastToken(node),
-            startLine = node.loc.start.line,
-            endLine = lastToken.loc.end.line;
-
-        return startLine === endLine;
-    }
-
-    /**
      * Returns the VariableDeclarator based on the current node
      * if not present then return null
      * @param {ASTNode} node node to examine
@@ -361,6 +281,80 @@ module.exports = function(context) {
         return varNode &&
             varNode.parent.loc.start.line === node.loc.start.line &&
             varNode.parent.declarations.length > 1;
+    }
+
+    /**
+     * Check indent for function block content
+     * @param {ASTNode} node node to examine
+     * @returns {void}
+     */
+    function checkIndentInFunctionBlock(node) {
+
+        // Search first caller in chain.
+        // Ex.:
+        //
+        // Models <- Identifier
+        //   .User
+        //   .find()
+        //   .exec(function() {
+        //   // function body
+        // });
+        //
+        // Looks for 'Models'
+        var calleeNode = node.parent; // FunctionExpression
+        var indent;
+
+        if (calleeNode.parent &&
+            (calleeNode.parent.type === "Property" ||
+            calleeNode.parent.type === "ArrayExpression")) {
+            // If function is part of array or object, comma can be put at left
+            indent = getNodeIndent(calleeNode, false, true);
+        } else {
+            // If function is standalone, simple calculate indent
+            indent = getNodeIndent(calleeNode);
+        }
+
+        // If part of a multi-line chain call, indent one more to differentiate:
+        // var a = new P()
+        //     .done(function (result) {
+        //         doUsefulStuff(result);
+        //     });
+        var calleeLoc = calleeNode.type === "CallExpression" && calleeNode.callee.loc;
+        if (calleeLoc &&
+            calleeLoc.end.line - calleeLoc.start.line === 1 &&
+            calleeLoc.end.line < node.loc.start.line
+        ) {
+            indent += indentSize;
+        }
+
+        // function body indent should be indent + indent size
+        indent += indentSize;
+
+        // check if the node is inside a variable
+        var parentVarNode = getVariableDeclaratorNode(node);
+        if (parentVarNode && isNodeInVarOnTop(node, parentVarNode)) {
+            indent += indentSize * options.VariableDeclarator[parentVarNode.parent.kind];
+        }
+
+        if (node.body.length > 0) {
+            checkNodesIndent(node.body, indent);
+        }
+
+        checkLastNodeLineIndent(node, indent - indentSize);
+    }
+
+
+    /**
+     * Checks if the given node starts and ends on the same line
+     * @param {ASTNode} node The node to check
+     * @returns {boolean} Whether or not the block starts and ends on the same line.
+     */
+    function isSingleLineNode(node) {
+        var lastToken = context.getLastToken(node),
+            startLine = node.loc.start.line,
+            endLine = lastToken.loc.end.line;
+
+        return startLine === endLine;
     }
 
     /**

--- a/lib/rules/no-unused-vars.js
+++ b/lib/rules/no-unused-vars.js
@@ -111,10 +111,10 @@ module.exports = function(context) {
      */
     function isUsedVariable(variable, references) {
         var functionNodes = variable.defs.filter(function(def) {
-            return def.type === "FunctionName";
-        }).map(function(def) {
-            return def.node;
-        }),
+                return def.type === "FunctionName";
+            }).map(function(def) {
+                return def.node;
+            }),
             isFunctionDefinition = functionNodes.length > 0;
 
         return references.some(function(ref) {

--- a/tests/fixtures/rules/indent/indent-invalid-fixture-1.js
+++ b/tests/fixtures/rules/indent/indent-invalid-fixture-1.js
@@ -411,7 +411,7 @@ a(); // ->
 }
 
 a( "very very long multi line" +
-   "with weird indentation", function() {
+    "with weird indentation", function() {
   b();
 a(); // ->
     c(); // <-

--- a/tests/fixtures/rules/indent/indent-valid-fixture-1.js
+++ b/tests/fixtures/rules/indent/indent-valid-fixture-1.js
@@ -301,10 +301,10 @@ a
   }]);
 
 var a = function() {
-      a++;
-      b++; // <-
-      c++; // <-
-    },
+    a++;
+    b++; // <-
+    c++; // <-
+  },
   b;
 
 var b = [
@@ -358,8 +358,8 @@ if (array.some(function(){
 }
 
 var a = b.c(function() {
-      d++;
-    }),
+    d++;
+  }),
   e;
 
 switch (true) {
@@ -411,11 +411,11 @@ if( "very very long multi line" +
 }
 
 a( "very very long multi line" +
-   "with weird indentation", function() {
-  b();
-  a(); // ->
-  c(); // <-
-});
+    "with weird indentation", function() {
+      b();
+      a(); // ->
+      c(); // <-
+    });
 
 a = function(content, dom) {
   b();
@@ -424,14 +424,14 @@ a = function(content, dom) {
 };
 
 a = function(content, dom) {
-      b();
-      c(); // <-
-      d(); // ->
-    };
+  b();
+  c(); // <-
+  d(); // ->
+};
 
 a = function(content, dom) {
-    b(); // ->
-  };
+  b(); // ->
+};
 
 a = function(content, dom) {
   b(); // ->
@@ -439,9 +439,9 @@ a = function(content, dom) {
 
 a('This is a terribly long description youll ' +
   'have to read', function () {
-  b();
-  c();
-});
+    b();
+    c();
+  });
 
 if (
   array.some(function(){

--- a/tests/lib/rules/indent.js
+++ b/tests/lib/rules/indent.js
@@ -54,6 +54,37 @@ ruleTester.run("indent", rule, {
     valid: [
         {
             code:
+            "function test() {\n" +
+            "  return client.signUp(email, PASSWORD, { preVerified: true })\n" +
+            "    .then(function (result) {\n" +
+            "      // hi\n" +
+            "    })\n" +
+            "    .then(function () {\n" +
+            "      return FunctionalHelpers.clearBrowserState(self, {\n" +
+            "        contentServer: true,\n" +
+            "        contentServer1: true\n" +
+            "      });\n" +
+            "    });\n" +
+            "}",
+            options: [2]
+        },
+        {
+            code:
+            "function test() {\n" +
+            "    return client.signUp(email, PASSWORD, { preVerified: true })\n" +
+            "    .then(function (result) {\n" +
+            "        var x = 1;\n" +
+            "        var y = 1;\n" +
+            "    }, function(err){\n" +
+            "        var o = 1 - 2;\n" +
+            "        var y = 1 - 2;\n" +
+            "        return true;\n" +
+            "    });\n" +
+            "}",
+            options: [4]
+        },
+        {
+            code:
             "// hi",
             options: [2, {"VariableDeclarator": 1, "SwitchCase": 1}]
         },
@@ -910,8 +941,9 @@ ruleTester.run("indent", rule, {
                 [189, 2, 0, "VariableDeclaration"],
                 [193, 6, 4, "ExpressionStatement"],
                 [195, 6, 8, "ExpressionStatement"],
-                [305, 6, 4, "ExpressionStatement"],
-                [306, 6, 8, "ExpressionStatement"],
+                [304, 4, 6, "ExpressionStatement"],
+                [306, 4, 8, "ExpressionStatement"],
+                [307, 2, 4, "BlockStatement"],
                 [308, 2, 4, "VariableDeclarator"],
                 [311, 4, 6, "Identifier"],
                 [312, 4, 6, "Identifier"],
@@ -941,6 +973,8 @@ ruleTester.run("indent", rule, {
                 [349, 2, 0, "ExpressionStatement"],
                 [355, 2, 0, "ExpressionStatement"],
                 [357, 2, 4, "ExpressionStatement"],
+                [361, 4, 6, "ExpressionStatement"],
+                [362, 2, 4, "BlockStatement"],
                 [363, 2, 4, "VariableDeclarator"],
                 [368, 2, 0, "SwitchCase"],
                 [370, 2, 4, "SwitchCase"],
@@ -952,15 +986,23 @@ ruleTester.run("indent", rule, {
                 [392, 2, 4, "ExpressionStatement"],
                 [409, 2, 0, "ExpressionStatement"],
                 [410, 2, 4, "ExpressionStatement"],
-                [416, 2, 0, "ExpressionStatement"],
-                [417, 2, 4, "ExpressionStatement"],
+                [415, 6, 2, "ExpressionStatement"],
+                [416, 6, 0, "ExpressionStatement"],
+                [417, 6, 4, "ExpressionStatement"],
+                [418, 4, 0, "BlockStatement"],
                 [422, 2, 4, "ExpressionStatement"],
                 [423, 2, 0, "ExpressionStatement"],
-                [428, 6, 8, "ExpressionStatement"],
-                [429, 6, 4, "ExpressionStatement"],
-                [434, 2, 4, "BlockStatement"],
+                [427, 2, 6, "ExpressionStatement"],
+                [428, 2, 8, "ExpressionStatement"],
+                [429, 2, 4, "ExpressionStatement"],
+                [430, 0, 4, "BlockStatement"],
+                [433, 2, 4, "ExpressionStatement"],
+                [434, 0, 4, "BlockStatement"],
                 [437, 2, 0, "ExpressionStatement"],
                 [438, 0, 4, "BlockStatement"],
+                [442, 4, 2, "ExpressionStatement"],
+                [443, 4, 2, "ExpressionStatement"],
+                [444, 2, 0, "BlockStatement"],
                 [451, 2, 0, "ExpressionStatement"],
                 [453, 2, 4, "ExpressionStatement"],
                 [499, 6, 8, "BlockStatement"],

